consensus/obcpbft/broadcast.go

```diff
@@ -67,35 +73,68 @@ func (b *broadcaster) Wait() {
 	b.closed.Wait()
 }
 
-func (b *broadcaster) unicastOne(msg *pb.Message, dest uint64, wait chan bool) {
+func (b *broadcaster) drainerSend(dest uint64, send *sendRequest, printedValidatorNotFound bool) bool {
 	defer func() {
+		send.done <- struct{}{}
 		b.closed.Done()
-		wait <- true
 	}()
+	h, err := getValidatorHandle(dest)
+	if err != nil {
+		if !printedValidatorNotFound {
+			logger.Warningf("could not get handle for replica %d", dest)
+		}
+		time.Sleep(time.Second)
+		return true
+	}
 
-	select {
-	case <-b.closedCh:
-		return
-	default:
+	if printedValidatorNotFound {
+		logger.Infof("Found handle for replica %d", dest)
+		printedValidatorNotFound = false
 	}
 
-	h, err := getValidatorHandle(dest)
+	err = b.comm.Unicast(send.msg, h)
 	if err != nil {
-		logger.Warningf("could not get handle for replica %d", dest)
-	} else {
-		err = b.comm.Unicast(msg, h)
+		logger.Warningf("could not send to replica %d: %v", dest, err)
 	}
-	if err != nil {
-		logger.Debugf("could not send to replica %d: %v", dest, err)
+
+	return false
+}
+
+func (b *broadcaster) drainer(dest uint64) {
+	printedValidatorNotFound := false
+
+	for {
 		select {
 		case <-b.closedCh:
-			return
-		case <-time.After(1 * time.Second):
-			return
+			for {
+				// Drain the message channel to free calling waiters before we shut down
+				select {
+				case send := <-b.msgChans[dest]:
+					send.done <- struct{}{}
+					b.closed.Done()
+				default:
+					return
+				}
+			}
+		case send := <-b.msgChans[dest]:
+			printedValidatorNotFound = b.drainerSend(dest, send, printedValidatorNotFound)
 		}
 	}
 }
 
+func (b *broadcaster) unicastOne(msg *pb.Message, dest uint64, wait chan struct{}) {
+	select {
+	case b.msgChans[dest] <- &sendRequest{
+		msg:  msg,
+		done: wait,
+	}:
+	default:
```

Comment:
Yes, I considered that briefly, I was thinking maybe instead we could only count 'successful' sends towards the total count, so instead of replying with that struct, switch it back to a bool.

Maybe the logic could then be "we wait for 2f+1 successes, or after n seconds, 2f+1 replies of success or failure". What do you think?